### PR TITLE
Black 606

### PR DESCRIPTION
**Pull request by saltysealion**
Introducing 3 new factories: SmoothPulseFactory, CustomPulseFactory & OptimalPulseFactory

Custom pulses allow a user to specify their pulse via segments
Optimal Pulses allow users to specify what pulse they would like Q-CTRL to optimize towards. 
Smooth pulses allow users to filter their pulses

I have merged Ryan's work with optional related factories in. Enabling invoking actions with or without their related models.

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -18,7 +18,7 @@ workflows:
           filters:
             tags:  # Tag filter required because the release job has a tag filter
               only: /.*/              
-      - q-ctrl/build:
+      - q-ctrl/build_and_publish_internally:
```
Comment (stabacco) [inline .circleci/config.yml]: weird to see these changes here, seem unrelated.

File: qctrl/factories/pulse_factory.py [hunk 2]
```diff
@@ -73,3 +73,26 @@ def fixed_modulus(self):
...
+@FACTORY_REGISTRY.register("CustomPulseFactory")
+class CustomPulseFactory(QctrlModelFactory):
+    """A custom pulse is a pulse whose segments are defined by the user.
+    """
+    class Meta:
+        """Pulse factory meta.
+        """
+        model = "pulses"
+
+    type="custom"
```
Comment (stabacco) [inline qctrl/factories/pulse_factory.py]: we should probably look into moving these into commons at some stage, so that we can be sure they are in sync across packages.

File: .circleci/config.yml (hunk 1 above)
Comment (saltysealion) [reply .circleci/config.yml]: Yup, never too sure what's going on with auto commits these days

File: qctrl/factories/pulse_factory.py (hunk 2 above)
Comment (saltysealion) [reply qctrl/factories/pulse_factory.py]: That's a good idea
